### PR TITLE
Improve waf

### PR DESCRIPTION
**Pull request by petya2164**
@mortenvp After some struggle with pytest, I managed to run your unit tests on most of our buildslaves: http://buildbot.steinwurf.dk/powerconsole?project=waf
As you can see, I created the `test` function in the wscript to invoke pytest in a completely standalone fashion. I also tried tox, but it doesn't work properly on Windows and it is an overkill for the build system, since we already test on many different environments. Of course, you can run tox locally if you like, but it should not be mandatory.
We still have some issues with the configure step on CentOS and Windows, and the logs are not helping much:
http://buildbot.steinwurf.dk/builders/waf-centos7_64-cxx_gxx52_x64-py27_x64/builds/6/steps/configure/logs/stdio
http://buildbot.steinwurf.dk/builders/waf-win7_64-cxx_msvc14_x64-py27_x64/builds/6/steps/configure/logs/stdio
I guess we should talk about the next steps when you have some time ;)

### CODE REVIEW

**Comment by petya2164** [comment]
@mortenvp I finally got an all-green build with the currently enabled tests. I had to abandon the self-building approach, I added some basic logic in buildbot.py to build and test in a standalone way.
Here's one annoying issue when using waf for testing waf itself:
If you deployed a new waf binary to the project root, and it fails for some reason, then you cannot rebuild waf with this new waf to fix the issue (you have to revert manually). But if we build waf without waf, then we won't face that problem.
For now, I propose to move the root wscript to `test/test_self_build` to run the self-building test via pytest in an isolated environment. When we get close to the actual release and this works reliably on all platforms, then we can consider going back to self-building (and we might also kill the submodules at that point).

**Comment by petya2164** [comment]
@mortenvp I created the separated self_build test as I explained above and the buildslaves are able to run this test with one caveat: the `bundle_dependencies` folder is not created automatically and git clone fails as it is invoked with a non-existent working directory. I just had to manually create the folder in the test: https://github.com/steinwurf/waf/blob/improve-waf/test/self_build/test_self_build.py#L20
For a real fix, we should ensure that the specified `bundle_path` always exists with `os.makedirs`, but this variable is used at multiple places and I don't know where to add this. Can you try to fix that?

I made some smaller changes and cleaned up some obsolete code. 
I have a few thoughts/suggestions for future tasks:
1. If something goes wrong, then it is really tedious to find the problem in the logs (especially if this happens on a buildslave), so we should always print the full traceback for the last exception instead of hiding it in the logs. Whenever we fail with ctx.fatal, we should print a user-friendly error message and then everything about the exception. I did this here: https://github.com/steinwurf/waf/blob/improve-waf/src/wurf/dependency_manager.py#L62 and this complete output helped me a lot! If we do this everywhere, then opening the log should be rarely necessary.
2. I realized that we don't need to find the actual path to the git binary, we just invoke `git` and the system knows where it is. With this simplification, the git commands in the output look much nicer (especially on Windows). In some situations (e.g. standalone archive), waf should be able to run on a system where git is not even installed. If all the dependencies are present, then we don't have to call git for anything.
3. The code of the 3 GitResolvers is very similar and I would like to unify these to avoid duplication and reduce confusion. I would also like to return to the old folder structure within `bundle_dependencies`: one folder for each dependency, and inside we should have a subfolder for `master`, the tags and the checkouts. The current folder structure looks inconsistent and confusing for me with that double hash :(

Since everything is green now, I think this is a good time for you to merge and we can continue from there.
After merging, you can build and test with these commands:
```
python buildbot.py build
python buildbot.py run_tests
```

P.S. You can see those little end-of-line whitespace fixes in the diff, I think it would be good to enable `trim_trailing_whitespace` in your editor if possible ;)

**Comment by petya2164** [comment]
@mortenvp I removed the submodules and managed to make everything green with the self-build approach (using the latest waf binary from your branch). I also switched to mock 1.0.1.

I think this won't get any simpler, it should be safe for you to merge all these changes. Actually, we might also get rid of shutilwhich, see my comments above about git ;)

File: src/wurf/waf_build_context.py [hunk 1]
```diff
@@ -3,22 +3,15 @@
...
-        # right before starting a build.
-        #
-        # https://github.com/waf-project/waf/blob/master/waflib/Build.py#L269
-        #
-        #  Which is more or less what we want to do. Here we recurse into to the
-        #  dependencies build functions allowing them to also hook up
-        #  user-defined build functions.
-        waf_resolve_context.recurse_dependencies(self)
-        super(WafBuildContext, self).pre_build()
+    def pre_recurse(self, node):
```
Comment (mortenvp) [inline src/wurf/waf_build_context.py]: I'm not sure these change are not needed. We just have to fix our waf-tools and the dependencies wscripts.

File: src/wurf/waf_configuration_context.py [hunk 2]
```diff
@@ -3,18 +3,19 @@
...
 
-    def execute(self):
+    def pre_recurse(self, node):
 
-        super(WurfConfigurationContext, self).execute()
+        super(WafConfigurationContext, self).pre_recurse(node)
 
-        # Call configure in all dependencies
-        waf_resolve_context.recurse_dependencies(self)
+        # Call configure() in all dependencies before executing configure()
```
Comment (mortenvp) [inline src/wurf/waf_configuration_context.py]: I think we should not make this tied to our waf-tools .. it if can be avoided.. therefore I would like to try to fix the waf-tools before trying to fix stuff here.

File: wscript [hunk 3]
```diff
@@ -175,7 +178,21 @@ def _pytest(bld):
     separator = ';' if sys.platform == 'win32' else ':'
     bld_env.env.update({'PYTHONPATH': separator.join(python_path)})
 
-    bld(rule='python -m pytest test',
+    # Remove the previously created temp folder
```
Comment (mortenvp) [inline wscript]: Could you add some comment why this is needed.

File: buildbot.py [hunk 4]
```diff
@@ -15,16 +15,20 @@ def run_command(args, cwd=None, env=None):
...
 
+
 def build(properties):
     command = [sys.executable, 'waf', 'build', '-v']
     run_command(command)
 
+
 def run_tests(properties):
-    pass
+    command = [sys.executable, 'waf', '-v', '--run_tests']
```
Comment (mortenvp) [inline buildbot.py]: Should we not instead make the `--skip_tests` the default. I personally normally want to run the tests after building.

Comment (petya2164) [inline buildbot.py]: --run_tests is the norm in all our projects, and this also creates a separate run_tests step on the build system. I like that very much!

Comment (petya2164) [inline buildbot.py]: To be perfectly honest, I find it extremely weird that "python waf build" runs any tests. That should be a different step.

File: wscript (hunk 3 above)
Comment (petya2164) [inline wscript]: Done

**Pull request merged**